### ISSUE TITLE
Making sure heavy atoms come first in the sorting algorithm

### ISSUE THREAD

**Issue by nickvandewiele** [ReactionMechanismGenerator/RMG-Py#452]
Currently, the sorting algorithm recursively computes the connectivity values of _all_ the vertices in the graph, regardless of whether the atom is hydrogen or not.

```python
connectivityValues = self.__update([len(vertex.edges) for vertex in self.vertices], 0)
```

This elegant solution, however, has the disadvantage that hydrogen atoms might end up in a position in the array of atoms, prior to heavy atoms.

e.g.:

```
1  C u0 p0 c0 {2,S} {3,S} {4,S} {9,S}
2  C u0 p0 c0 {1,S} {3,S} {11,S} {12,S}
3  C u1 p0 c0 {1,S} {2,S} {13,S}
4  C u0 p0 c0 {1,S} {5,D} {6,S}
5  C u0 p0 c0 {4,D} {7,S} {14,S}
6  C u0 p0 c0 {4,S} {8,D} {15,S}
7  C u0 p0 c0 {5,S} {10,D} {16,S}
8  C u0 p0 c0 {6,D} {10,S} {17,S}
9  H u0 p0 c0 {1,S}
10 C u0 p0 c0 {7,D} {8,S} {18,S}
11 H u0 p0 c0 {2,S}
12 H u0 p0 c0 {2,S}
13 H u0 p0 c0 {3,S}
14 H u0 p0 c0 {5,S}
15 H u0 p0 c0 {6,S}
16 H u0 p0 c0 {7,S}
17 H u0 p0 c0 {8,S}
18 H u0 p0 c0 {10,S}
```

We should look for an implementation in which heavy atoms always end up _before_ hydrogen atoms.

**Comment by nickvandewiele** [comment]
An additional complication that I just discovered is that the _isomorphism_ algorithm fails when two graphs are sorted in a different way.

Currently, we have two sorting algorithms that differ in implementation:
- `rmgpy.molecule.molecule.Molecule.sortAtoms()` sorts Molecule objects based on Atom attributes
- `rmgpy.molecule.graph.Graph.sortVertices()` sorts Graph objects based on Vertex attributes (in fact: only 1: the connectivity value)

Within `rmgpy.molecule.vf2.VF2.isomorphism(graph1, graph2)` we sort the graphs using `sortVertices()`. But the vertices will only be when the vertices have not been sorted yet (`vertex.sortingLabel < 0`). If the vertices of 1 graph are sorted using `sortAtoms()` while the other graph's vertices are sorted using `sortVertices`, then a false negative is produced.

**Comment by nickvandewiele** [comment]
The reason why I made a new sorting algorithm (`Molecule.sortAtoms()`) is because the original algorithm, that solely uses vertex connectivity values, sometimes fails in distinguishing non-identical atoms. The latter is a prerequisite to correctly parse/generate unique identifiers to/from a Molecule.

**Comment by rwest** [comment]
If canonical ordering were a prerequisite to VF2 then 1) we wouldn't have had it working and 2) the isomorphism would be trivial. I thought the only requirement was that the ordering was stable during the isomorphism algorithm. Though I suppose we may have introduced optimizations relying on the same sorting algorithm being applied. 
This is from memory/speculation.

**Comment by nickvandewiele** [comment]
https://github.com/ReactionMechanismGenerator/RMG-Py/blob/master/rmgpy/molecule/vf2.pyx#L220

```python
    cdef bint feasible(self, Vertex vertex1, Vertex vertex2) except -2:
            ...
            if vertex1.connectivity != vertex2.connectivity: return False
```

this line generates trouble if the connectivity values of the 2 vertices are derived from distinct sorting algorithms.

**Comment by nickvandewiele** [comment]
so, indeed, it is an optimization shortcut that is the culprit here.

**Comment by rwest** [comment]
So it's not the sorting that's the problem, it's the different definitions of connectivity values.

**Comment by nickvandewiele** [comment]
My fix would include 2 things:
1) removing that line that uses connectivity values to decide whether two atoms are potential matches.
2) ensuring that atoms are sorted in the same way when calling the isomorphism algorithm, since the speedup this generates for the isomorphism checks is though to be significant.

An implementation of the latter fix could be done in various ways:
- the least efficient way would be to reset the connectivity values each time an isomorphism check is done.
- Storing the way the atoms were sorted. Currently the -1 flag for `Vertex.sortingLabel` only stores whether it _is_ sorted, not _how_ it is sorted. Moreover, the variable `Vertex.sortingLabel` is set to the index of the atom once the atoms are sorted.

**Comment by nickvandewiele** [comment]
The presence of 2 sorting algorithms that reset the order of the atoms is really messing up a lot of things.

The only reason why I need the customized sorting algorithm is for augmented InChIs, so I am inclined to move all that functionality to the parser module, and only use it there.

I would replace all calls to `sortAtoms` outside the parser module by `sortVertices` calls instead to restore the original flow again.

**Comment by rwest** [comment]
If you need a canonical order for augmented InChIs, you could do worse than using the AuxInfo from the InChI API to get their sorting order. https://github.com/ReactionMechanismGenerator/RMG-Py/pull/443#issuecomment-135916799

**Comment by nickvandewiele** [comment]
This is fixed by #456. Future improvements will be made based on the comments in this issue.
